extensions/namespace-lookup/src/main/java/io/druid/server/namespace/cache/NamespaceExtractionCacheManager.java

```diff
@@ -336,150 +336,72 @@ private boolean removeNamespaceLocalMetadata(final NamespaceImplData implDatum)
-                  }
-                }
-              },
-              0,
-              updateMs,
-              TimeUnit.MILLISECONDS
-          );
-        } else {
-          final Map<String, String> cache = getCacheMap(cacheId);
-          final Callable<String> runnable = factory.getCachePopulator(namespace, null, cache);
-          final ListenableFuture<String> futureWithString = listeningScheduledExecutorService.schedule(
-              new Callable<String>()
-              {
-                @Override
-                public String call() throws Exception
-                {
-                  startLatch.await(); // wait for "election" to leadership or cancellation
-                  tasksStarted.incrementAndGet();
-                  return runnable.call();
-                }
-              },
-              0,
-              TimeUnit.MILLISECONDS
-          );
-          Futures.addCallback(
-              futureWithString, new FutureCallback<String>()
-              {
-                @Override
-                public void onSuccess(String result)
-                {
-                  try {
-                    postRunnable.run();
-                  }
-                  catch (RuntimeException e) {
-                    delete(cacheId);
-                    throw e;
-                  }
-                  // Must have been set in order to make it here
-                  if (implDatum.get().enabled.get()) {
-                    lastVersion.put(namespaceName, result);
-                  }
-                }
 
-                @Override
-                public void onFailure(Throwable t)
-                {
-                  // NOOP
-                }
-              }
-          );
-          future = futureWithString;
+    final Runnable command = new Runnable()
+    {
+      @Override
+      public void run()
+      {
+        try {
+          startLatch.await(); // wait for "election" to leadership or cancellation
+          if (!Thread.currentThread().isInterrupted()) {
+            final Map<String, String> cache = getCacheMap(cacheId);
+            final String preVersion = lastVersion.get(namespaceName);
+            final Callable<String> runnable = factory.getCachePopulator(namespace, preVersion, cache);
+
+            tasksStarted.incrementAndGet();
+            final String newVersion = runnable.call();
+            if (preVersion != null && preVersion.equals(newVersion)) {
+              throw new IllegalStateException("Already exists");
+            }
+            if (newVersion != null) {
+              lastVersion.put(namespaceName, newVersion);
+            }
+            postRunnable.run();
+            log.debug("Namespace [%s] successfully updated", namespaceName);
+          }
         }
-      }
-      catch (Exception e) {
-        if (future != null) {
-          if (!future.isDone() && !future.cancel(true)) {
-            log.warn("Could not cancel future for [%s]", namespaceName);
+        catch (Throwable t) {
```

Comment:
InterruptedException will be throw to caller. Is that should be handled in exception handler?